### PR TITLE
Refactor/libraries

### PR DESCRIPTION
**Pull request by provokateurin**
Closes https://github.com/provokateurin/nextcloud-neon/issues/433

### CODE REVIEW

File: packages/neon/neon/lib/utils.dart [hunk 1]
```diff
@@ -0,0 +1,12 @@
+export 'package:neon/src/utils/app_implementation.dart';
```
Comment (Leptopoda) [inline packages/neon/neon/lib/utils.dart:1]: I never liked this being a util.

It should rather be a model or interface but I don't know how much it makes sense :(

File: packages/neon/neon/lib/blocs.dart
Comment (Leptopoda) [inline packages/neon/neon/lib/blocs.dart:1]: Should this also export bloc?
How much does it make sense to have them separated once the AccountsBloc isn't exported anymore?

File: packages/neon/neon/lib/utils.dart [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+export 'package:neon/src/utils/branding.dart';
+export 'package:neon/src/utils/confirmation_dialog.dart';
+export 'package:neon/src/utils/exceptions.dart';
+export 'package:neon/src/utils/hex_color.dart';
+export 'package:neon/src/utils/nextcloud_app_specific_options.dart';
+export 'package:neon/src/utils/rename_dialog.dart';
+export 'package:neon/src/utils/request_manager.dart';
+export 'package:neon/src/utils/sort_box_builder.dart';
+export 'package:neon/src/utils/sort_box_order_option_values.dart';
+export 'package:neon/src/utils/storage.dart';
```
Comment (Leptopoda) [inline packages/neon/neon/lib/utils.dart:11]: moved to settings?

**Review by Leptopoda — CHANGES_REQUESTED** [review]
Some random thoughts. They may not all make sense.

- I think that all libs should add the `library` directive with a small little doc comment.
- All private stuff should be annotated with `@internal` so we don't accidentally export it (directly or indirectly). This might be useful for internal widgets and the blocs.

**Comment by Leptopoda** [comment]
I just noticed that neon is directly exposing types defined in nextcloud.

If dynamite or the specs would change some naming this would not only result in breaking changes to nextcloud but also in neon.
We should abstract away all nextcloud types that we expose into their own typedefs. Internally we can still use the normal types.

File: packages/neon/neon/lib/interfaces.dart [hunk 3]
```diff
@@ -0,0 +1 @@
+export 'package:neon/src/interfaces/notifications.dart';
```
[the remark this answers is not in the document]
Comment (provokateurin) [reply packages/neon/neon/lib/interfaces.dart:1]: We could do that, but for me models is something to model concrete data and not to model some abstract interface that gets implemented elsewhere

**Comment by provokateurin** [comment]
I agree with you on most points. Moving to models makes sense :+1:

File: packages/neon/neon/lib/blocs.dart
Comment (provokateurin) [reply packages/neon/neon/lib/blocs.dart:1]: Yeah let's expose them together as blocs.

**Comment by provokateurin** [comment]
@Leptopoda  Should we also export file_icons and nextcloud through the neon package?

File: packages/neon/neon/lib/src/widgets/drawer_destination.dart
Comment (Leptopoda) [inline packages/neon/neon/lib/src/widgets/drawer_destination.dart:1]: should be exported as it's gonna be needed for the adaptive layout.

Or should I just expoprt it then?

File: packages/neon/neon/lib/src/utils/stream_listenable.dart
Comment (Leptopoda) [inline packages/neon/neon/lib/src/utils/stream_listenable.dart:1]: Could be exported (doesn't need to)

File: packages/neon/neon/lib/settings.dart
Comment (Leptopoda) [inline packages/neon/neon/lib/settings.dart:1]: Could we structure the settings folder in a way similar to all other packages?

/settings
|-> widgets
|-> models
| -> ...

it's currently all over the place

**Review by Leptopoda — COMMENTED** [review]
I'd really like to have internal annotations on things that we don't (want to ever) export.
[0001-neon-add-internal-annotations.patch.txt](https://github.com/provokateurin/nextcloud-neon/files/11929134/0001-neon-add-internal-annotations.patch.txt)

File: packages/neon/neon/lib/src/widgets/drawer_destination.dart
Comment (provokateurin) [reply packages/neon/neon/lib/src/widgets/drawer_destination.dart:1]: Yeah just export it when needed. I only added exports for things that are used outside the neon package right now

File: packages/neon/neon/lib/src/utils/stream_listenable.dart
Comment (provokateurin) [reply packages/neon/neon/lib/src/utils/stream_listenable.dart:1]: Just export it when you need it

**Comment by Leptopoda** [comment]
> @Leptopoda Should we also export file_icons and nextcloud through the neon package?

Yeah this would probably make thing easier for devs.
But I don't know where :(
The neon library is currently not needed for apps (like files, notes ...) and only needed the `app` itself.
We'd need to come up with a better place for this.

**Comment by provokateurin** [comment]
> I'd really like to have internal annotations on things that we don't (want to ever) export.

I looked at your patch and I don't agree with all of it. For example `NeonUserAvatar` will be needed for Talk as we want to display the user avatars there.

**Comment by provokateurin** [comment]
>  We'd need to come up with a better place for this.

Ok let's keep this in mind and do it at some later point

**Comment by Leptopoda** [comment]
> I looked at your patch and I don't agree with all of it. For example NeonUserAvatar will be needed for Talk as we want to display the user avatars there.

I added it to every major internal class. I didn't know you are currently only exporting used things.
I'm fine with not everything having them but things like the theme should probably have them.

**Review by Leptopoda — APPROVED** [review]
LGTM

**Pull request merged**